File: data/cpp/db/data/json/JsonWriter.cpp
/*
 * Copyright (c) 2007-2009 Digital Bazaar, Inc. All rights reserved.
 */
#include "db/data/json/JsonWriter.h"

#include "db/rt/DynamicObjectIterator.h"
#include "db/rt/Exception.h"
#include "db/io/BufferedOutputStream.h"
#include "db/io/OStreamOutputStream.h"

#include <iostream>
#include <sstream>

using namespace std;
using namespace db::data;
using namespace db::data::json;
using namespace db::io;
using namespace db::rt;

JsonWriter::JsonWriter(bool strict)
{
   mStrict = strict;
   // Initialize to compact representation
   setCompact(true);
   setIndentation(0, 3);
}

JsonWriter::~JsonWriter()
{
}

bool JsonWriter::writeIndentation(OutputStream* os, int level)
{
   bool rval = true;
   int indent = mCompact ? 0 : (level * mIndentSpaces);
   
   // write out indentation
   if(indent > 0)
   {
      char temp[indent];
      memset(temp, ' ', indent);
      rval = os->write(temp, indent);
   }
   
   return rval;
}

bool JsonWriter::write(DynamicObject& dyno, OutputStream* os, int level)
{
   bool rval = true;

   if(level < 0)
   {
      level = mIndentLevel;
   }

   if(dyno.isNull())
   {
      rval = os->write("null", 4);
   }
   else
   {
      switch(dyno->getType())
      {
         case String:
         {
            const char* temp = dyno->getString();
            size_t length = strlen(temp);
            
            // Note: UTF-8 has a maximum of 6-bytes per character when
            // encoded in json format ("/u1234")
            string encoded;
            encoded.reserve(length);
            encoded.push_back('"');
            char unicode[6];
            for(size_t i = 0; i < length; i++)
            {
               unsigned char c = temp[i];
               if((c >= 0x5d /* && c <= 0x10FFFF */) ||
                  (c >= 0x23 && c <= 0x5B) ||
                  (c == 0x21) ||
                  (c == 0x20))
               {
                  // TODO: check this handles UTF-* properly
                  encoded.push_back(c);
               }
               else
               {
                  encoded.push_back('\\');
                  switch(c)
                  {
                     case '"': /* 0x22 */
                     case '\\': /* 0x5C */
                     // '/' is in the RFC but not required to be escaped
                     //case '/': /* 0x2F */
                        encoded.push_back(c);
                        break;
                     case '\b': /* 0x08 */
                        encoded.push_back('b');
                        break;
                     case '\f': /* 0x0C */
                        encoded.push_back('f');
                        break;
                     case '\n': /* 0x0A */
                        encoded.push_back('n');
                        break;
                     case '\r': /* 0x0D */
                        encoded.push_back('r');
                        break;
                     case '\t': /* 0x09 */
                        encoded.push_back('t');
                        break;
                     default:
                        // produces "u01af" (4 digits of 0-filled hex)
                        snprintf(unicode, 6, "u%04x", c);
                        encoded.append(unicode, 5);
                        break;
                  }
               }
            }
            
            // end string serialization and write encoded string
            encoded.push_back('"');
            rval = os->write(encoded.c_str(), encoded.length());
            break;
         }
         case Boolean:
         case Int32:
         case UInt32:
         case Int64:
         case UInt64:
         case Double:
         {
            const char* temp = dyno->getString();
            rval = os->write(temp, strlen(temp));
            break;
         }
         case Map:
         {
            // start map serialization
            rval = (mCompact) ? os->write("{", 1) : os->write("{\n", 2);
            
            // serialize each map member
            DynamicObjectIterator i = dyno.getIterator();
            while(rval && i->hasNext())
            {
               DynamicObject& next = i->next();
               
               // serialize indentation and start serializing member name
               if((rval =
                  writeIndentation(os, level + 1) &&
                  os->write("\"", 1) &&
                  os->write(i->getName(), strlen(i->getName()))))
               {
                  // end serializing member name, serialize member value
                  rval = ((mCompact) ?
                     os->write("\":", 2) : os->write("\" : ", 4)) &&
                     write(next, os, level + 1);
                  
                  // serialize delimiter if appropriate
                  if(rval && i->hasNext())
                  {
                     rval = os->write(",", 1);
                  }
                  
                  // add formatting if appropriate
                  if(rval && !mCompact)
                  {
                     rval = os->write("\n", 1);
                  }
               }
            }
            
            // end map serialization
            rval = writeIndentation(os, level) && os->write("}", 1);
            break;
         }
         case Array:
         {
            // start array serialization
            rval = (mCompact) ? os->write("[", 1) : os->write("[\n", 2);
            
            // serialize each array element
            DynamicObjectIterator i = dyno.getIterator();
            while(rval && i->hasNext())
            {
               // serialize indentation and array value
               rval =
                  writeIndentation(os, level + 1) &&
                  write(i->next(), os, level + 1);
               
               // serialize delimiter if appropriate
               if(rval && i->hasNext())
               {
                  rval = os->write(",", 1);
               }
               
               // add formatting if appropriate
               if(rval && !mCompact)
               {
                  rval = os->write("\n", 1);
               }
            }
            
            // end array serialization
            rval = writeIndentation(os, level) && os->write("]", 1);
            break;
         }
      }
   }
   
   return rval;
}

bool JsonWriter::write(DynamicObject& dyno, OutputStream* os)
{
   bool rval = true;
   
   if(mStrict)
   {
      rval = !dyno.isNull();
      
      DynamicObjectType type;
      if(rval)
      {
         type = dyno->getType();
      }
   
      if(!rval || !(type == Map || type == Array))
      {
         ExceptionRef e = new Exception(
            "No JSON top-level Map or Array found.",
            "db.data.json.JsonWriter.InvalidJson");
         Exception::setLast(e, false);
         rval = false;
      }
   }
   
   if(rval)
   {
      ByteBuffer b(1024);
      BufferedOutputStream bos(&b, os);
      rval = write(dyno, &bos, mIndentLevel) && bos.flush();
   }
   
   return rval;
}

void JsonWriter::setCompact(bool compact)
{
   mCompact = compact;
}

void JsonWriter::setIndentation(int level, int spaces)
{
   mIndentLevel = level;
   mIndentSpaces = spaces;
}

bool JsonWriter::writeToOStream(
   DynamicObject& dyno, ostream& stream, bool compact, bool strict)
{
   OStreamOutputStream os(&stream);
   JsonWriter jw(strict);
   jw.setCompact(compact);
   if(!compact)
   {
      jw.setIndentation(0, 3);
   }
   return jw.write(dyno, &os);
}

std::string JsonWriter::writeToString(
   DynamicObject& dyno, bool compact, bool strict)
{
   string rval;
   
   ostringstream oss;
   if(writeToOStream(dyno, oss, compact, strict))
   {
      rval = oss.str();
   }
   
   return rval;
}

bool JsonWriter::writeToStdOut(
   DynamicObject& dyno, bool compact, bool strict)
{
   bool rval;
   if((rval = writeToOStream(dyno, cout, compact, strict)))
   {
      cout << endl;
      cout.flush();
   }
   return rval;
}

File: data/java/src/com/db/data/JsonWriter.java
/*
 * Copyright (c) 2008 Digital Bazaar, Inc.  All rights reserved.
 */
package com.db.data;

import java.io.BufferedOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.Arrays;

/**
 * A JsonWriter provides an interface for serializing objects to
 * JSON (JavaScript Object Notation) (RFC 4627).
 * 
 * A JsonWriter writes out a whole object at once and can be used again.
 * The compact setting should be used to minimize extra whitespace when not
 * needed.
 * 
 * @author David I. Lehn
 * @author Dave Longley
 */
public class JsonWriter
{
   /**
    * Compact mode to minimize whitespace.
    */
   protected boolean mCompact;
   
   /**
    * The starting indentation level. 
    */
   protected int mIndentLevel;
   
   /**
    * The number of spaces per indentation level.
    */
   protected int mIndentSpaces;
   
   /**
    * Creates a new JsonWriter.
    */
   public JsonWriter()
   {
      // Initialize to compact representation
      setCompact(true);
      setIndentation(0, 3);
   }
   
   /**
    * Writes out indentation.  None if in compact mode.
    * 
    * @param os the OutputStream to write to.
    * @param level indentation level.
    * 
    * @exception IOException thrown if an IO error occurs. 
    */
   protected void writeIndentation(OutputStream os, int level)
      throws IOException
   {
      int indent = mCompact ? 0 : (level * mIndentSpaces);
      
      // write out indentation
      if(indent > 0)
      {
         byte[] temp = new byte[indent];
         byte c = ' ';
         Arrays.fill(temp, c);
         os.write(temp, 0, indent);
      }
   }
   
   /**
    * Recursively serializes an object to JSON using the passed DynamicObject.
    * 
    * @param dyno the DynamicObject to serialize.
    * @param os the OutputStream to write the JSON to.
    * @param level current level of indentation (-1 to initialize with default).
    * 
    * @exception IOException thrown if an IO error occurs.
    */
   protected void write(DynamicObject dyno, OutputStream os, int level)
      throws IOException
   {
      if(level < 0)
      {
         level = mIndentLevel;
      }
      
      if(dyno == null)
      {
         os.write(new String("null").getBytes("ASCII"));
      }
      else
      {
         switch(dyno.getType())
         {
            case String:
               {
                  String temp = dyno.getString();
                  int length = temp.length();
                  
                  // UTF-8 has a maximum of 7-bytes per character when
                  // encoded in json format
                  StringBuilder encoded = new StringBuilder(length * 7 + 2);
                  encoded.append('"');
                  for(int i = 0; i < length; i++)
                  {
                     char c = temp.charAt(i);
                     if((c >= 0x5d /* && c <= 0x10FFFF */) ||
                        (c >= 0x23 && c <= 0x5B) ||
                        (c == 0x21) ||
                        (c == 0x20))
                     {
                        // TODO: check this handles UTF-* properly
                        encoded.append(c);
                     }
                     else
                     {
                        encoded.append('\\');
                        switch(c)
                        {
                           case '"': /* 0x22 */
                           case '\\': /* 0x5C */
                              encoded.append(c);
                              break;
                           // '/' is in the RFC but not required to be escaped
                           //case '/': /* 0x2F */
                           //   encoded[n++] = '/';
                           //   break;
                           case '\b': /* 0x08 */
                              encoded.append('b');
                              break;
                           case '\f': /* 0x0C */
                              encoded.append('f');
                              break;
                           case '\n': /* 0x0A */
                              encoded.append('n');
                              break;
                           case '\r': /* 0x0D */
                              encoded.append('r');
                              break;
                           case '\t': /* 0x09 */
                              encoded.append('t');
                              break;
                           default:
                              encoded.append(String.format("u%04x", (int)c));
                              break;
                        }
                     }
                  }
                  
                  // end string serialization and write encoded string
                  encoded.append('"');
                  os.write(encoded.toString().getBytes("ASCII"));
               }
               break;
            case Boolean:
            case Number:
               {
                  os.write(dyno.getString().getBytes("ASCII"));
               }
               break;
            case Map:
               {
                  // start map serialization
                  if(mCompact)
                  {
                     byte b = '{';
                     os.write(b);
                  }
                  else
                  {
                     byte[] b = {'{','\n'};
                     os.write(b);
                  }
                  
                  // serialize each map member
                  byte b;
                  byte[] compactEndName = {'"',':'};
                  byte[] endName = {'"',' ',':',' '};
                  DynamicObjectIterator i = dyno.getIterator();
                  while(i.hasNext())
                  {
                     DynamicObject next = i.next();
                     
                     // serialize indentation and start serializing member name
                     writeIndentation(os, level + 1);
                     b = '"';
                     os.write(b);
                     os.write(i.getName().getBytes("UTF-8"));
                     
                     // end serializing member name, serialize member value
                     if(mCompact)
                     {
                        os.write(compactEndName);
                     }
                     else
                     {
                        os.write(endName);
                     }
                     
                     // write object
                     write(next, os, level + 1);
                     
                     // serialize delimiter if appropriate
                     if(i.hasNext())
                     {
                        b = ',';
                        os.write(b);
                     }
                     
                     // add formatting if appropriate
                     if(!mCompact)
                     {
                        b = '\n';
                        os.write(b);
                     }
                  }
                  
                  // end map serialization
                  writeIndentation(os, level);
                  b = '}';
                  os.write(b);
               }
               break;
            case Array:
               {
                  // start array serialization
                  if(mCompact)
                  {
                     byte b = '[';
                     os.write(b);
                  }
                  else
                  {
                     byte[] b = {'[','\n'};
                     os.write(b);
                  }
                  
                  // serialize each array element
                  byte b;
                  DynamicObjectIterator i = dyno.getIterator();
                  while(i.hasNext())
                  {
                     // serialize indentation and array value
                     writeIndentation(os, level + 1);
                     write(i.next(), os, level + 1);
                     
                     // serialize delimiter if appropriate
                     if(i.hasNext())
                     {
                        b = ',';
                        os.write(b);
                     }
                     
                     // add formatting if appropriate
                     if(!mCompact)
                     {
                        b = '\n';
                        os.write(b);
                     }
                  }
                  
                  // end array serialization
                  writeIndentation(os, level);
                  b = ']';
                  os.write(b);
               }
               break;
         }
      }
   }
   
   /**
    * Serializes an object to JSON using the passed DynamicObject.
    * 
    * @param dyno the DynamicObject to serialize.
    * @param os the OutputStream to write the JSON to.
    * 
    * @exception IOException thrown if an IO error occurs.
    */
   public void write(DynamicObject dyno, OutputStream os)
      throws IOException
   {
      BufferedOutputStream bos = new BufferedOutputStream(os);
      write(dyno, bos, mIndentLevel);
      bos.flush();
   }
   
   /**
    * Sets the starting indentation level and the number of spaces
    * per indentation level.
    * 
    * @param level the starting indentation level.
    * @param spaces the number of spaces per indentation level.
    */
   public void setIndentation(int level, int spaces)
   {
      mIndentLevel = level;
      mIndentSpaces = spaces;
   }
   
   /**
    * Sets the writer to use compact mode and not output unneeded whitespace.
    * 
    * @param compact true to minimize whitespace, false not to.
    */
   public void setCompact(boolean compact)
   {
      mCompact = compact;
   }
}
